File: gb/wikipedia.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-


import urllib
import urllib2
import re
 

def encodetitle(title):
    title = title.replace(' ', '_')
    title = urllib.quote(title.encode('utf8'))
    return title


def getpage(title, encode=True):
    opener = urllib2.build_opener()
    opener.addheaders = [('User-agent', 'Mozilla/5.0')]
    if encode:
        title = encodetitle(title)
    url = 'http://en.wikipedia.org/w/index.php?title=' + title + '&action=raw';
    try:
        infile = opener.open(url)
    except:
        print "ERROR - getpage(): couldn't open url"
        return ''
    s = infile.read()

    return ''.join(s)


def page2sections(page):
    sections = {}
    cur_section = ''
    cur_text = ''

    lines = page.split('\n')
    for l in lines:
        if (len(l) > 4) and (l[:2] == '=='):
            if len(cur_text) > 0:
                sections[cur_section] = cur_text
            section = l.strip(' =')
            section = section.lower()
            cur_section = section
            cur_text = ''
        else:
            cur_text += l + '\n'

    if len(cur_text) > 0:
        sections[cur_section] = cur_text

    return sections


def text_and_or_link(s):
    text = ''
    link = ''
    matches = re.findall('\[\[([^\]]*)\]\]', s)
    if len(matches) > 0:
        parts = matches[0].split('|')
        link = parts[0].strip()
        if len(parts) > 1:
            text = parts[1].strip()
        else:
            text = link
    else:
        text = s.strip()

    return text, link


def br_list(s):
    result = []
    items = s.split('<br />')
    for i in items:
        result.append(i.strip())
    return result


def get_image_html(title):
    opener = urllib2.build_opener()
    opener.addheaders = [('User-agent', 'Mozilla/5.0')]
    title = encodetitle(title)
    url = 'http://en.wikipedia.org/wiki/File:' + title;
    infile = opener.open(url)
    s = infile.read()

    return ''.join(s)


def get_image_url(name):
    url = ''
    html = get_image_html(name)
    m = re.findall('upload.wikimedia.org/wikipedia/en/([^/]*)/([^/]*)', html)
    if len(m) > 0:
        url = 'http://upload.wikimedia.org/wikipedia/en/%s/%s/%s' % (m[0][0], m[0][1], encodetitle(name))
    return url

File: brain-generators/src/main/scala/Wikipedia.scala
package com.graphbrain.brain-generators

import java.net.URL
import scala.io.Source
import scala.util.matching.Regex
import scala.xml._
import java.io.BufferedReader;
import java.io.InputStreamReader
import scala.collection.mutable.ListBuffer
import scala.collection.mutable.HashMap

object Wikipedia {
	val wikipediaBaseURL = "http://en.wikipedia.org/w/index.php?title="
  	val enWikiPath="en.wikipedia.org/wiki/"
  	val raw="&action=raw"
  	val ListRegex = """List(s?)\sof\s""".r
	 val WikiLinkRegex="""\[\[([^\]]*)\]\]""".r
  	val CategoryRegex="""category\:.*""".r
  	val IrregularTitleRegex=""".*\:.*""".r
    val pipe="""\|""".r
    

  def getPage(title:String, baseURL:String=wikipediaBaseURL, getRaw:Boolean=true):String={
    	var pageURL = baseURL +title.replace(" ", "_")
    	if(getRaw)
    	{
      		pageURL+=raw
    	}
    	val page=new URL(pageURL);
    	val reader=new BufferedReader(new InputStreamReader(page.openStream()));
    	var input="";
    	while(true)
    	{
      		val line=reader.readLine()
      		line match{
        		case a:String => input+=line+"\n"; 
        		case _ => return input;
      		}

    	}
    	return input;
    }

  def getLinks(content:String):List[String]={
    val foundLinks = WikiLinkRegex.findAllIn(content)
    var links:List[String]=List()
    while(foundLinks.hasNext)
    {
      val link=foundLinks.next().replaceAll("""(\[\[)""", "").replaceAll("""\]\]""", "").replaceAll("""\|.*""", "").replaceAll("""#""", "")
      links=link::links
    }
    return links.reverse

  }

  def text_and_or_link(s:String):(String, String)={
  	val regex ="""\[\[([^\]]*)\]\]""".r
  	val matches = regex.findAllIn(s)
  	var text = ""
  	var link = ""
  	if(matches.hasNext)
  	{
  		val parts = matches.next.split("|");
  		val link = parts(0).trim;
  		if(parts.length>1)
  		{
  			text = parts(0).trim
  		}
  		else{
  			text = link
  		}
  	}
  	else
  		text = s.trim
  	return(text, link)
  }

  def br_list(s:String):Array[String]=
  {

    return s.split("<br>")
  }
    

  def getInfobox(content:String, fields:List[String]):HashMap[String, Array[String]]=
  {
    var infoItems:HashMap[String, Array[String]] = new HashMap
    val lines = content.split('\n')
    for(line <- lines)
    {

      if(line.contains('='))
      {
        
        val splitItems = line.split('=')
        val fieldname=pipe.replaceAllIn(splitItems(0), "").trim
        if(fields.contains(fieldname))
        {
          val values = splitItems(1)
          infoItems.update(fieldname, values.split("<br>"))  
        }
        

      } 
      
    }
    return infoItems
    
  }

  def page2sections(content:String):HashMap[String, String]=
  {
    var sections:HashMap[String, String] = new HashMap()
    var cur_section = ""
    var cur_text = ""

    val lines = content.split("\n")
    for(l <- lines)
    {
    	if(l.length>4&&l.substring(0,2)=="==")
    	{
    		if(cur_text.length>0)
    		{
    			sections.update(cur_section, cur_text)
    		}
    		val section = l.replaceAll("=", "").toLowerCase
    		cur_section = section;
    		cur_text="";

    	}
    	else{
    		cur_text += l + "\n";
    	}
    	if(cur_text.length>0)
    	{
    		sections.update(cur_section, cur_text)
    	}
    }
    return sections
 }

 def page2sectionsLists(content:String):HashMap[String, List[String]]=
  {
    var sections:HashMap[String, List[String]] = new HashMap()
    var cur_section = ""
    var cur_text = ""

    val lines = content.split("\n")
    for(l <- lines)
    {
    	if(l.length>4&&l.substring(0,2)=="==")
    	{
    		if(cur_text.length>0)
    		{
    			sections.update(cur_section, cur_text::sections.getOrElse(cur_section, Nil))
    		}
    		val section = l.replaceAll("=", "").toLowerCase
    		cur_section = section;
    		cur_text="";

    	}
    	else{
    		cur_text += l + "\n";
    	}
    	if(cur_text.length>0)
    	{
    		sections.update(cur_section, cur_text::sections.getOrElse(cur_section, Nil))
    	}
    }
    return sections
 }
}